### PR TITLE
Improve docs of leaderelection configuration

### PR DESCRIPTION
**Pull request by ChrsMark**
Follow up of https://github.com/elastic/beats/pull/20512.

### CODE REVIEW

**Comment by masci** [comment]
Following up https://github.com/elastic/beats/pull/20512#issuecomment-674016999

> I would say that the current approach is smoother for every case. I don't think we should remove the Deployment completely by now since in many cases it can be really helpful when it comes to scaling where someone may want to collect cluster-wide metrics from big clusters.

I'm not sure we should chase complex use cases but focus on onboarding instead. 

With the current approach, users don't benefit from the simplified configuration provided by leader election out of the box: on the contrary, they have to learn about specifics of the feature through the comments and change the manifest in several points to simplify the setup, which is a bit of a paradox.

If I'm running a big, complex cluster requiring to fine-tune observability, chances are I won't even use the manifests as they are in the repo because the default resources provided to the standalone pods won't fit any possible case.

To sum up, I think with the current setup we're not providing a turnkey solution to anybody, while with leader election we have the chance to build a good user story to simplify onboarding.

**Comment by ChrsMark** [comment]
@exekias @jsoriano what do you think about @masci 's proposal of completely removing the old proposed manifests that include the Deployment and keep only the new approach with the leader election configuration within the Daemonset?

I'm ok with this, however I'm only concerned about users that already use Deployment manifests (breaking change ?).

**Comment by jsoriano** [comment]
> @exekias @jsoriano what do you think about @masci 's proposal of completely removing the old proposed manifests that include the Deployment and keep only the new approach with the leader election configuration within the Daemonset?

Maybe we have to check how big a deployment needs to be to require a dedicated pod. There are two main reasons why this may be needed:
* Collection of cluster-level metrics require significantly more resources, and configuring these resources in the DaemonSet will make most of the pods to reserve much more resources than they need.
* Performance problems with leader election caused by having two many candidates.

If we are more or less sure that for most of the cases this is not going to be a problem, then I am ok with keeping the new approach only.

> I'm ok with this, however I'm only concerned about users that already use Deployment manifests (breaking change ?).

We can keep a dummy deployment with the same name and zero replicas to replace the previous one avoiding a breaking change.

**Comment by ChrsMark** [comment]
> > @exekias @jsoriano what do you think about @masci 's proposal of completely removing the old proposed manifests that include the Deployment and keep only the new approach with the leader election configuration within the Daemonset?
> 
> Maybe we have to check how big a deployment needs to be to require a dedicated pod. There are two main reasons why this may be needed:
> 
> * Collection of cluster-level metrics require significantly more resources, and configuring these resources in the DaemonSet will make most of the pods to reserve much more resources than they need.
> * Performance problems with leader election caused by having two many candidates.
> 
> If we are more or less sure that for most of the cases this is not going to be a problem, then I am ok with keeping the new approach only.
> 

My feeling is that this kind of performance issues will hit users with real big clusters and if they have to fine tune their setup to resolve this issues it shouldn't be a big deal to evaluate the solution of the singleton Pod using a deployment. For most of our users keeping only the new approach from now on would improve their onboarding experience.

> > I'm ok with this, however I'm only concerned about users that already use Deployment manifests (breaking change ?).
> 
> We can keep a dummy deployment with the same name and zero replicas to replace the previous one avoiding a breaking change.

I think that having a dummy Deployment spec inside our manifests making things more complicated for users that are not so familiar with the whole setup. On the other hand I'm not sure if this can be considered a breaking change since docs are linked to the versions of the manifests and only old users that might want adopt the new approach might end up with "orphan" Deployment Pods on their clusters however this is something that not so unusual and it's not that bad.


In this, I'm leaning towards keeping only the new approach.
@masci @jsoriano @exekias let me know what you think.

**Comment by exekias** [comment]
It would sound like the general case would benefit from going only for the DaemonSet, and then we can **document** how to switch to a cluster-wide Deployment for big clusters.

This is not really a breaking change, as it won't break existing deployments, we are just changing how the new ones will happen.

Br

**Comment by jsoriano** [comment]
Ok, should we mention something in the docs in case someone upgrades by using the new manifests?

> "orphan" Deployment Pods on their clusters however this is something that not so unusual and it's not that bad.

Well, these orphan deployments will be collecting duplicated metrics, with an old version.

**Comment by ChrsMark** [comment]
> Ok, should we mention something in the docs in case someone upgrades by using the new manifests?
> 

New docs will only point to new manifests and only refer to the new approach. And wouldn't be an overkill to mention this in the old versions' docs?

> > "orphan" Deployment Pods on their clusters however this is something that not so unusual and it's not that bad.
> 
> Well, these orphan deployments will be collecting duplicated metrics, with an old version.

**Comment by jsoriano** [comment]
> New docs will only point to new manifests and only refer to the new approach. And wouldn't be an overkill to mention this in the old versions' docs?

I meant to mention only something like "if you deployed these manifests before 7.10, you may need to manually remove the deployment". But we are not telling anything about upgrades so far, so maybe it is ok to don't say anything and users just change the versions in the manifests they are using.

File: metricbeat/docs/running-on-kubernetes.asciidoc [hunk 1]
```diff
@@ -17,18 +17,18 @@ endif::[]
...
 metrics, Docker stats, and metrics from all the services running on top of
 Kubernetes.
 
-* As a single {beatname_uc} instance created using a https://kubernetes.io/docs/concepts/workloads/controllers/Deployment/[Deployment].
+In addition one Pod of the Daemonset will hold a leader lock each time which makes it responsible for
+handling cluster wide monitoring.
 This instance is used to retrieve metrics that are unique for the whole
 cluster, such as Kubernetes events or
 https://github.com/kubernetes/kube-state-metrics[kube-state-metrics].
+You canfind more information about leader election configuration options at <<configuration-autodiscover>>.
```
Comment (jsoriano) [inline metricbeat/docs/running-on-kubernetes.asciidoc:31]:
```suggestion
You can find more information about leader election configuration options at <<configuration-autodiscover>>.
```

File: metricbeat/docs/running-on-kubernetes.asciidoc [hunk 2]
```diff
@@ -200,41 +200,11 @@ Metrics should start flowing to Elasticsearch.
...
-kubectl create -f metricbeat-kubernetes.yaml
-------------------------------------------------
+==== Deploying Metricbeat to collect cluster level metrics for big clusters
+
+In some cases, users might want to monitor really bog Kubernetes clusters. In such cases,
+the Pod that will be collecting cluster level metrics might face performance issues due to
+resources limitations. In this case users might consider avoid using leader election configuration
+for the Daemonset and instead deploy a singleton Metribceat instance using a
+https://github.com/elastic/beats/blob/7.9/deploy/kubernetes/metricbeat/metricbeat-deployment.yaml[Deployment] with the
+proper https://github.com/elastic/beats/blob/7.9/deploy/kubernetes/metricbeat/metricbeat-deployment-configmap.yaml[configuration].
```
Comment (jsoriano) [inline metricbeat/docs/running-on-kubernetes.asciidoc:210]: This is linking to older branches, we may easily forget to apply changes there that could be needed for future versions. Would it make sense to somehow copy the important parts of these manifests to some doc file? Or we just remove these links and leave to the user the definition of their configuration?

File: metricbeat/docs/running-on-kubernetes.asciidoc [hunk 3]
```diff
@@ -200,41 +200,11 @@ Metrics should start flowing to Elasticsearch.
...
-the Daemonset on the cluster with the leader election provider enabled
-in order to collect cluster-wide metrics:
-["source", "sh", subs="attributes"]
-------------------------------------------------
-curl -L -O https://raw.githubusercontent.com/elastic/beats/{branch}/deploy/kubernetes/metricbeat-kubernetes.yaml
-kubectl create -f metricbeat-kubernetes.yaml
-------------------------------------------------
+==== Deploying Metricbeat to collect cluster level metrics for big clusters
+
+In some cases, users might want to monitor really bog Kubernetes clusters. In such cases,
```
Comment (jsoriano) [inline metricbeat/docs/running-on-kubernetes.asciidoc:205]:
```suggestion
In some cases, users might want to monitor really big Kubernetes clusters. In such cases,
```

File: metricbeat/docs/running-on-kubernetes.asciidoc (hunk 2 above)
Comment (ChrsMark) [reply metricbeat/docs/running-on-kubernetes.asciidoc:210]: 🤔 My intention here is to point to the latest version that had been supporting the Deployment approach just for reference. I don't see any reason of updating this however do you see any case I might miss?

Comment (jsoriano) [reply metricbeat/docs/running-on-kubernetes.asciidoc:213]: If this is going to be the documentation for big clusters I think we should keep it updated. Imagine in 7.10/7.11, someone with a big cluster sees this section, and they use the 7.9 deployment linked here. They may end up using an incorrect version, and they may also be missing some configuration change that we haven't backported.

Comment (ChrsMark) [reply metricbeat/docs/running-on-kubernetes.asciidoc]: 👍  gotcha, thanks!

**Review by jsoriano — APPROVED** [review]

File: metricbeat/docs/running-on-kubernetes.asciidoc [hunk 4]
```diff
@@ -17,18 +17,21 @@ endif::[]
...
-
-* As a https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/[DaemonSet]
+You deploy {beatname_uc} as a https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/[DaemonSet]
 to ensure that there's a running instance on each node of the cluster. These
 instances are used to retrieve most metrics from the host, such as system
 metrics, Docker stats, and metrics from all the services running on top of
 Kubernetes.
 
-* As a single {beatname_uc} instance created using a https://kubernetes.io/docs/concepts/workloads/controllers/Deployment/[Deployment].
+In addition one Pod of the Daemonset will hold a leader lock each time which makes it responsible for
```
Comment (masci) [inline metricbeat/docs/running-on-kubernetes.asciidoc:26]:
```suggestion
In addition, one of the Pods in the DaemonSet will constantly hold a _leader lock_ which makes it responsible for
```

File: metricbeat/docs/running-on-kubernetes.asciidoc [hunk 5]
```diff
@@ -17,18 +17,21 @@ endif::[]
...
+handling cluster wide monitoring.
```
Comment (masci) [inline metricbeat/docs/running-on-kubernetes.asciidoc:27]:
```suggestion
handling cluster-wide monitoring.
```

File: metricbeat/docs/running-on-kubernetes.asciidoc [hunk 6]
```diff
@@ -17,18 +17,21 @@ endif::[]
...
 This instance is used to retrieve metrics that are unique for the whole
 cluster, such as Kubernetes events or
 https://github.com/kubernetes/kube-state-metrics[kube-state-metrics].
+You can find more information about leader election configuration options at <<configuration-autodiscover>>.
+
+Note: If you are upgrading from older versions of these manifests please make sure that there are no redundant parts
```
Comment (masci) [inline metricbeat/docs/running-on-kubernetes.asciidoc:33]:
```suggestion
Note: If you are upgrading from older versions, please make sure there are no redundant parts
```

File: metricbeat/docs/running-on-kubernetes.asciidoc [hunk 7]
```diff
@@ -17,18 +17,21 @@ endif::[]
...
+as left-overs from previous versions. Deployment specification and its ConfigMaps might be the case.
```
Comment (masci) [inline metricbeat/docs/running-on-kubernetes.asciidoc:34]:
```suggestion
as left-overs from the old manifests. Deployment specification and its ConfigMaps might be the case.
```

File: metricbeat/docs/running-on-kubernetes.asciidoc [hunk 8]
```diff
@@ -200,41 +203,11 @@ Metrics should start flowing to Elasticsearch.
...
-      - config:
-          - module: kubernetes
-            hosts: ["kube-state-metrics:8080"]
-            period: 10s
-            add_metadata: true
-            metricsets:
-              - state_node
------
-Users can find more info about the `unique` and `identifier` options at <<configuration-autodiscover>>.
+==== Deploying Metricbeat to collect cluster level metrics for big clusters
```
Comment (masci) [inline metricbeat/docs/running-on-kubernetes.asciidoc:206]:
```suggestion
==== Deploying Metricbeat to collect cluster-level metrics in large clusters
```

File: metricbeat/docs/running-on-kubernetes.asciidoc [hunk 9]
```diff
@@ -200,41 +203,11 @@ Metrics should start flowing to Elasticsearch.
...
+
+In some cases, users might want to monitor really big Kubernetes clusters. In such cases,
```
Comment (masci) [inline metricbeat/docs/running-on-kubernetes.asciidoc:208]:
```suggestion
The size and the number of nodes in a Kubernetes cluster can be fairly large at times, and in such cases
```

File: metricbeat/docs/running-on-kubernetes.asciidoc [hunk 10]
```diff
@@ -200,41 +203,11 @@ Metrics should start flowing to Elasticsearch.
...
+the Pod that will be collecting cluster level metrics might face performance issues due to
+resources limitations. In this case users might consider avoid using leader election configuration
```
Comment (masci) [inline metricbeat/docs/running-on-kubernetes.asciidoc:210]:
```suggestion
resources limitations. In this case users might consider to avoid using the leader election strategy
```

File: metricbeat/docs/running-on-kubernetes.asciidoc [hunk 11]
```diff
@@ -200,41 +203,11 @@ Metrics should start flowing to Elasticsearch.
...
+for the Daemonset and instead deploy a singleton Metribceat instance using a Deployment with the
```
Comment (masci) [inline metricbeat/docs/running-on-kubernetes.asciidoc:211]:
```suggestion
and instead run a dedicated, standalone Metribceat instance using a Deployment in addition to the DaemonSet.
```

File: metricbeat/docs/running-on-kubernetes.asciidoc [hunk 12]
```diff
@@ -200,41 +203,11 @@ Metrics should start flowing to Elasticsearch.
...
+proper configuration.
```
Comment (masci) [inline metricbeat/docs/running-on-kubernetes.asciidoc:212]:
```suggestion
```

**Review by masci — APPROVED** [review]
Left few wording suggestions for the docs but this LGTM!
